src/Simulation/QCTraceSimulator/DepthCounter.cs

```diff
@@ -27,15 +27,24 @@ private class OperationCallRecord
             public QubitTimeMetrics[] InputQubitMetrics;
         }
 
-        StatisticsCollector<CallGraphEdge> stats;
-        Stack<OperationCallRecord> operationCallStack;
-        QubitAvailabilityTimeTracker qubitAvailabilityTime = new QubitAvailabilityTimeTracker(
-            initialCapacity: 128, // Reasonable number to preallocate.
-            defaultAvailabilityTime: 0.0);
+        private readonly bool optimizeDepth = false;
+        private readonly StatisticsCollector<CallGraphEdge> stats;
+        private readonly Stack<OperationCallRecord> operationCallStack;
+
+        // Used when OptimizeDepth = false. We track availability time for each underlying qubits identified by id.
+        private readonly QubitAvailabilityTimeTracker qubitAvailabilityTime;
+
+        // Used when OptimizeDepth = true. We reuse qubits based on busy intervals.
```

Comment:
It's mildly concerning that some of the fields of the class are only used depending on a value of another field. Are they at least valid with some meaningful defaults or must every use of them be conditioned on `optimizeDepth` setting? If the latter, maintenance of this code will be likely prone to introducing bugs :(

Reply:
Yes, that's true. I use either qubitAvailabilityTime or the other two fields. Unused fields are left at default value of null. I'm not sure there's a better option short of creating two different depth counters for two different alternatives.

Reply:
All problems can be solved by another layer of indirection! :D You could add an interface DepthOptimizationPolicy with methods for each place in the code when you currently have `if (OprimizeDepth)`, then provide two concrete implementations... Your call whether it's worth doing something like this.

Reply:
It's on the same level as having two separate classes. And it doesn't bring clarity. So I'd like to keep it as is.